scripts/service.py

```diff
@@ -0,0 +1,286 @@
+        create_fifo = False
+        if os.path.exists(self.pipePath):
+            if not stat.S_ISFIFO(os.stat(self.pipePath).st_mode):
+                # as this is a specific file in our specific dir,
+                # assume it is used exclusively, so the current none pipe
+                # stuff is considered garbage and try to delete that here.
+                if os.path.isfile(self.pipePath):
+                    os.remove(self.pipePath)  # remove the file
+                    self.logInfo("removed file:" + self.pipePath)
+                elif os.path.isdir(self.pipePath):
+                    shutil.rmtree(self.pipePath)  # remove dir and all contains
+                    self.logInfo("removed dir:" + self.pipePath)
+                else:
+                    self.logErr("Unexpected file node at:" + self.pipePath)
+                    raise OSError(errno.EEXIST, "Unexpected file node exists", self.pipePath)
+
+                create_fifo = True
+        else:
+                create_fifo = True
+
+        if create_fifo:
+            try:
+                os.mkfifo(self.pipePath)
+            except OSError as err:
+                if err.errno != errno.EEXIST:
+                    self.loggErr("Error encountered while creating pipe")
+                    raise
+
+    def _closeFIFO(self):
+        if self.fifo != None:
+            try:
+                os.close(self.fifo)
+            except OSError:
+                pass # just ignore
+
+            self.fifo = None
+                
+    def logInfo(self, logtxt):
+        logger.info(self.logPrefix + logtxt)
+        
+    def logErr(self, logtxt):
+        logger.error(self.logPrefix + logtxt)
+        
+    def work(self):
+        """ virtual function for server and client """
+        pass
+
+class ServerInstance(ServiceSkeleton):
+    """ server side implementation """
+
+    def __init__(self, name, script):
+        super(ServerInstance, self).__init__(name)
+        self.script = script
+        self.epollObj = select.epoll()
+        
+    def _openRead(self):
+        """ server side open named pipe to read """
+        try:
+            # here to have write permission also is necessary to keep one writer
+            # to avoid pipe 'broken' in case of other closure.
+            self.fifo = os.open(self.pipePath, os.O_RDWR | os.O_NONBLOCK)
+            self.epollObj.register(self.fifo, select.EPOLLIN)
+        except OSError as err:
+            self.logErr("open file read error:" + str(err))
+            raise
+
+    def _executeSub(self):
+        """ execute the passed script with subprocess """
+        self.logInfo("will invoke scripts:" + self.script)
+        try:
+            subprocess.Popen(self.script)
+        except OSError as err:
+            self.logErr("execute script error:" + str(err))
+            
+    def _readRequest(self):
+        """ read all that is written in the pipe, so in case of input storm,
+            just do work once.
+        """
```

Comment:
This is the right idea, but there is a condition in which this goes wrong.

The reason we are running this as a service is that the script it invokes for us may take too long (either because it has to do a lot of work or because it just takes too long to communicate the results to xapi - which might be xapi's fault but we've still got to deal with it). When you are off in _executeSub() you are not reading the pipe, which means in the event of a storm clients can fill the pipe and block. This is fine from the server's perspective (it will correctly empty the pipe and invoke _executeSub() again) but lots of blocking clients may cause problems elsewhere in udev, delaying other scripts, running out of workers etc.

I fear you will need threads to solve that one :-(

Reply:
@TimSmithCtx , not sure if I understand correctly. But although what's actually executed in the _executeSub() may takes long, the main thread will not wait till it finishes. So the gap between the main thread leaves the pipe not read and the next read is only forking the sub process, should it should be very quick, and will not causing too many clients awaits there. It is probably not as you think the main will wait till the _executeSub() finishes and then back to read again.

Reply:
I guess I read that wrong then. What stops more than one _executeSub() from running at once?

Reply:
Yes, the subprocess will only fork the child, and only return a stub to the child process, and we did not do any extra wait/check, so just after fork, the server returns, none blockingly.

In the situation of this usb scan case, my understanding is to not miss any lasted events. So if we just skip the invokation if there is one already, the latest change might be missing (supposing the current one is in some final clean up stage). 

But if this service is made general, there might be such case that only one active job is expected. So I added an option from the server side to define the behavior, as one child only, or multiple.

Reply:
For use case here,  there is no such issue. The child will only call xapi to scan usb. It is a simple cli call. The problem is the script might block, worst in several minutes. And I think we might also need to prepare for the worst case, a dead block in the child, as we do not know what's in the scripts. Still, I think this will be a case by case decision. If any child has risk in concurrent, I think it should also be addressed in the child itself, as we cannot control who will run the scripts and how. Also, making a server block is always the worse choice.

Reply:
@TimSmithCtx , For your mentioned implementation, you can start the server with '-u' option, so there is only one running child.

Reply:
Yes that's the case we want.

My point is I think we'll never want any other cases, and if we do the solution will more properly be done elsewhere, but it doesn't seem worth holding this up.

Reply:
@TimSmithCtx , I think in the case here, multiple child/invoke is a better solution. Let me try to explain once again. Think of one udev (device add) event fires. So when the server is processing e1, child is triggered and is processing something, here to call the command to invoke xapi to do usbscan. Unknowing the timing and sequence, the xapi could quickly get triggered and processed, before the child returns completely. Then at the moment, after xapi complete the scanning and before the child returns completely, another udev event fires. In this case, if we just skip the invoking because the child is not yet returns, the event (for example new device added) will be missed from processing, which is not good.

Reply:
If the read side simply blocks until the child has returned completely, then any and all events which occurred after it was started will be sitting in the pipe, ready to be coalesced into a single "go again" instruction and trigger another run. The pipe might be full, but we don't care because the writers don't block.

Reply:
That's a bug in the child and should be fixed there. It would be sufficient to log that we started it. Worst case the faulty service stops working.